### PR TITLE
[IA-1156] Move async cluster creation to back Leo

### PR DESCRIPTION
**Pull request by rtitle**
Rebased branch with the latest changes, including scalafmt. It all compiles, some unit tests still failing though. Working through it…

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala [hunk 1]
```diff
@@ -113,31 +121,41 @@ abstract class LeoRoutes(
...
+                          .unsafeToFuture()
                       }
                     } ~
                     delete {
                       complete {
-                        leonardoService.deleteCluster(userInfo, GoogleProject(googleProject), clusterName).map { _ =>
-                          StatusCodes.Accepted
-                        }
+                        leonardoService
+                          .deleteCluster(userInfo, GoogleProject(googleProject), clusterName)
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala:127]: fyi, this can be `.deleteCluster(userInfo, GoogleProject(googleProject), clusterName).as(StatusCodes.Accepted)`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 2]
```diff
@@ -333,6 +328,13 @@ trait ClusterComponent extends LeoComponent {
         unmarshalFullCluster(recs).headOption
       }
 
+    def getMinimalClusterById(id: Long): DBIO[Option[Cluster]] =
+      findByIdQuery(id).result.headOption.map { recOpt =>
+        recOpt map { rec =>
+          unmarshalCluster(rec, Seq.empty, List.empty, Map.empty, List.empty, List.empty, List.empty)
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: shall we create a new data type say `MinimalCluster` for this? Stubbing those fields out feels pretty weird to me. As caller of this function, I would expect all the data is from DB instead of some values being stubbed out

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 3]
```diff
@@ -106,10 +105,10 @@ case class UserJupyterExtensionConfig(nbExtensions: Map[String, String] = Map(),
...
 case class ClusterResource(value: String) extends ValueObject
 
-case class DataprocInfo(googleId: Option[UUID],
-                        operationName: Option[OperationName],
-                        stagingBucket: Option[GcsBucketName],
-                        hostIp: Option[IP])
+case class DataprocInfo(googleId: Option[UUID] = None,
+                        operationName: Option[OperationName] = None,
+                        stagingBucket: Option[GcsBucketName] = None,
+                        hostIp: Option[IP] = None)
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: are these fields really can be any combination of `Some` and `None`? Or shouldn't `googleId` at least exist?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 4]
```diff
@@ -169,19 +168,17 @@ object Cluster {
...
-             operation: Option[Operation] = None,
-             stagingBucket: Option[GcsBucketName] = None,
              clusterImages: Set[ClusterImage] = Set.empty): Cluster =
     Cluster(
       internalId = internalId,
       clusterName = clusterName,
       googleProject = googleProject,
       serviceAccountInfo = serviceAccountInfo,
-      dataprocInfo = DataprocInfo(operation.map(_.uuid), operation.map(_.name), stagingBucket, None),
+      dataprocInfo = DataprocInfo(),
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: It feels to be Cluster should have `dataprocInfo` field as an `Option`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala [hunk 5]
```diff
@@ -2,71 +2,56 @@ package org.broadinstitute.dsde.workbench.leonardo
...
-    } yield bucketName
+
+      readerAcl = NonEmptyList
+        .fromList(bucketSAs)
+        .map(readers => Map(StorageRole.ObjectViewer -> readers))
+        .getOrElse(Map.empty)
+      ownerAcl = Map(StorageRole.ObjectAdmin -> NonEmptyList.one(leoEntity))
+
+      _ <- google2StorageDAO.insertBucket(googleProject, bucketName)
+      _ <- google2StorageDAO.setIamPolicy(bucketName, readerAcl ++ ownerAcl)
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:53]: 👍 🎉

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala [hunk 6]
```diff
@@ -24,19 +39,390 @@ case class ClusterIamSetupException(googleProject: GoogleProject)
...
+
+      // Resize the cluster in Google
+      _ <- IO.fromFuture(
+        IO(gdDAO.resizeCluster(cluster.googleProject, cluster.clusterName, numWorkers, numPreemptibles))
+      )
+    } yield ()
+
+  def setMasterMachineType(cluster: Cluster, machineType: MachineType): IO[Unit] =
+    cluster.instances.toList.traverse_ { instance =>
+      instance.dataprocRole match {
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala]: use `.traverse` instead of pattern matching?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala [hunk 7]
```diff
@@ -24,19 +39,390 @@ case class ClusterIamSetupException(googleProject: GoogleProject)
...
+  def generateServiceAccountKey(googleProject: GoogleProject,
+                                serviceAccountEmailOpt: Option[WorkbenchEmail]): IO[Option[ServiceAccountKey]] =
+    serviceAccountEmailOpt.traverse { email =>
+      IO.fromFuture(IO(googleIamDAO.createServiceAccountKey(googleProject, email)))
+    }
+
+  def removeServiceAccountKey(googleProject: GoogleProject,
+                              serviceAccountEmailOpt: Option[WorkbenchEmail],
+                              serviceAccountKeyIdOpt: Option[ServiceAccountKeyId]): IO[Unit] =
+    (serviceAccountEmailOpt, serviceAccountKeyIdOpt).mapN {
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala:265]: maybe `parMapN`? since the two requests can run at the same time?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala [hunk 8]
```diff
@@ -24,19 +39,390 @@ case class ClusterIamSetupException(googleProject: GoogleProject)
...
+                                            initBucketName: GcsBucketName,
+                                            serviceAccountInfo: ServiceAccountInfo,
+                                            serviceAccountKeyOpt: Option[ServiceAccountKey]): IO[Unit] = {
+    // Clean up resources in Google
+    val deleteBucket = (for {
+      _ <- bucketHelper.deleteInitBucket(initBucketName)
+      _ <- log.info(
+        s"Successfully deleted init bucket ${initBucketName.value} for ${googleProject.value} / ${clusterName.value}"
+      )
+    } yield ()).recoverWith {
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala]: would it be a bit cleaner if you do` .attempt`, and then pattern match on the result?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/TemplateHelper.scala [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package org.broadinstitute.dsde.workbench.leonardo.util
+
+import java.io.File
+
+import cats.effect._
+import fs2._
+import org.broadinstitute.dsde.workbench.leonardo.config.ClusterResourcesConfig
+import org.broadinstitute.dsde.workbench.leonardo.model.ClusterResource
+
+object TemplateHelper {
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/TemplateHelper.scala:10]: nice 👍

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala [hunk 10]
```diff
@@ -121,15 +123,15 @@ class AuthProviderSpec
...
 
       // get status
-      val clusterStatus = leo.getActiveClusterDetails(userInfo, project, cluster1Name).futureValue
+      val clusterStatus = leo.getActiveClusterDetails(userInfo, project, cluster1Name).unsafeToFuture.futureValue
 
       cluster1 shouldEqual clusterStatus
 
       // list
-      val listResponse = leo.listClusters(userInfo, Map()).futureValue
+      val listResponse = leo.listClusters(userInfo, Map()).unsafeToFuture.futureValue
```
Comment (Qi77Qi) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala:155]: might as well just do `unsafeRunSync` in these places?

**Review by Qi77Qi — COMMENTED** [review]
Did a first pass 👍  Looks great

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 2 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Yeah, that's probably a good idea. Will look into it.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala [hunk 11]
```diff
@@ -113,31 +121,41 @@ abstract class LeoRoutes(
...
+                          .map { _ =>
+                            StatusCodes.Accepted
+                          }
+                          .unsafeToFuture()
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala]: shall we add something like https://github.com/broadinstitute/sam/blob/79417ba82faf5030ac98847940771fd941a801fb/src/main/scala/org/broadinstitute/dsde/workbench/sam/api/api.scala to get rid of those explicit `unsafeToFuture` calls? I don't feel strongly either way..but just fyi

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala (hunk 3 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: `googleId` is populated asynchronously so doesn't need to exist.

Yeah it probably makes more sense for `dataprocInfo` to be optional instead.

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala (hunk 10 above)
Comment (rtitle) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala:155]: I did `unsafeToFuture.futureValue` in all the tests. `.futureValue` is a ScalaTest thing which uses implicit `PatienceConfig`s, etc. Maybe we could make the tests use `IO` at some point but I didn't want to change them too much.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala (hunk 11 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala]: Ah cool.

**Comment by rtitle** [comment]
Thanks for the early feedback @Qi77Qi!

Unit tests finally passing, will check automation tests. Coveralls says:
```
coverage/coveralls — Coverage decreased (-1.4%) to 71.871%
```
which I find a bit suspicious because I removed a bunch of tests :)

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala (hunk 10 above)
Comment (Qi77Qi) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala:155]: ah make sense...I didn't think of `PatienceConfig`...

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala (hunk 7 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala:265]: Actually there's only 1 `IO` here -- the `mapN` is unwrapping 2 `Option`s

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 2 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: I'm thinking of maybe doing this in a separate PR…

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 12]
```diff
@@ -324,29 +303,22 @@ object ClusterInitValues {
...
+            contentSecurityPolicy: String): ClusterInitValues =
     ClusterInitValues(
-      googleProject.value,
-      clusterName.value,
-      stagingBucketName.value,
-      clusterImages.find(_.tool == Jupyter).map(_.dockerImage).getOrElse(""),
-      clusterImages.find(_.tool == RStudio).map(_.dockerImage).getOrElse(""),
+      cluster.googleProject.value,
+      cluster.clusterName.value,
+      cluster.dataprocInfo.map(_.stagingBucket.value).getOrElse(""),
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: not necessarily this PR, but I feel like this function should return an `Either[String, ClusterInitValues]` and in the case where these required fields (such as stagingBucket) is `None` for some reason, we can error out

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 13]
```diff
@@ -565,12 +540,15 @@ object LeonardoJsonSupport extends SprayJsonSupport with DefaultJsonProtocol {
...
-              fields.getOrElse("hostIp", JsNull).convertTo[Option[IP]]
-            ),
+            (fields.get("googleId"), fields.get("operationName"), fields.get("stagingBucket")).mapN {
+              (googleId, operationName, stagingBucket) =>
+                DataprocInfo(
+                  googleId.convertTo[UUID],
+                  operationName.convertTo[OperationName],
+                  stagingBucket.convertTo[GcsBucketName],
+                  fields.getOrElse("hostIp", JsNull).convertTo[Option[IP]]
+                )
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala:551]: semantic is a bit different between this and the original code...This code will return `None` if one of `googleId`, `operationName`, `stagingBucket` is `None`, vs the original logical seems to indicate these fields are not `None` or `Some` at the same time

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala (hunk 12 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: Yeah.. we currently just substitute with "" which may cause errors downstream. We could distinguish between required/optional fields and do some error checking up front.

Not going to do that in this PR, but good idea :)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala (hunk 13 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala:551]: Yeah this is a semantic change.. `DataprocInfo` is optional now, so I was saying to only construct it if `googleId`, `operationName`, and `stagingBucket` are present in the input json (which are required fields in `DataprocInfo`). I think this is valid based on the data model.. and actually, this json reader is never actually used since we never serialize input JSON to `Cluster` objects.

**Comment by rtitle** [comment]
fiab-start failed, looks unrelated to this PR. jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala (hunk 13 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala:551]: cool cool...That's way better than what it was before...Just wanted to call out the semantic change explicitly :)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala [hunk 14]
```diff
@@ -202,7 +202,7 @@ class ClusterMonitorActor(
         )
     } else {
       logger.info(
-        s"Cluster ${cluster.projectNameString} is not ready yet and has taken ${currTimeElapsed.toString} so far (Dataproc cluster status = $status, GCE instance statuses = ${instances
+        s"Cluster ${cluster.projectNameString} is not ready yet and has taken ${currTimeElapsed.millis.toSeconds} so far (Dataproc cluster status = $status, GCE instance statuses = ${instances
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: shall we add `seconds` after the variable? just so it's clear in the log what the unit is

**Review by Qi77Qi — APPROVED** [review]
:shipit:

**Comment by Qi77Qi** [comment]
some automation tests failed

**Comment by rtitle** [comment]
Thanks, yeah I see the bugs, will make a fix.

**Comment by rtitle** [comment]
Note: I also fixed https://broadworkbench.atlassian.net/browse/IA-1443 as part of this PR.

**Comment by rtitle** [comment]
fiab-start failed. jenkins retest

**Comment by rtitle** [comment]
fiab-start failed again? It looks unrelated to Leo. 

jenkins retest please

**Comment by rtitle** [comment]
All auto tests passed 🎉

File: http/src/main/resources/reference.conf [hunk 15]
```diff
@@ -106,11 +106,12 @@ monitor {
   pollPeriod = 15 seconds
   maxRetries = -1  # means retry forever
   recreateCluster = true
-  creatingTimeLimit = 1 hour
-  startingTimeLimit = 20 minutes
-  stoppingTimeLimit = 1 hour
-  deletingTimeLimit = 1 hour
-  updatingTimeLimit = 20 minutes
+  # Defines timeouts for cluster status transitions. If a status is not listed there is no timeout.
```
Comment (kyuksel) [inline http/src/main/resources/reference.conf:109]: Do we have documented somewhere what action is taken when a timeout is detected in each of the states below?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala [hunk 16]
```diff
@@ -175,40 +180,47 @@ class ClusterMonitorActor(
...
-
-      logger.info(
-        s"Detected that ${cluster.projectNameString} has been stuck in status $status too long. Failing it. Current timeout config: ${monitorConfig.monitorStatusTimeouts.toString}"
-      )
+    val currTimeElapsed: FiniteDuration = (System.currentTimeMillis() - startTime).millis
+    monitorConfig.monitorStatusTimeouts.get(cluster.status) match {
+      case Some(timeLimit) if currTimeElapsed > timeLimit =>
+        logger.info(s"Detected that ${cluster.projectNameString} has been stuck in status ${cluster.status} too long.")
+
+        // Take care not to Error out a cluster if it timed out in Starting status
```
Comment (kyuksel) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala:197]: Why are we no longer erroring out a cluster in this case?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala [hunk 17]
```diff
@@ -115,15 +105,14 @@ class ClusterMonitorSupervisor(
...
-  leonardoService: LeonardoService,
   clusterHelper: ClusterHelper
 )(implicit metrics: NewRelicMetrics[IO], clusterToolToToolDao: ClusterTool => ToolDAO[ClusterTool])
     extends Actor
     with Timers
     with LazyLogging {
   import context.dispatcher
 
-  var monitoredClusters: Set[Cluster] = Set.empty
+  var monitoredClusterIds: Set[Long] = Set.empty
```
Comment (kyuksel) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala:115]: I know this PR didn't introduce it but not having any implicit or explicit protection on this `var` against race conditions is a bit concerning, no? Doing something like in `ClusterDnsCache.hostToIpMapping` may be an easy remedy.

```scala
private val monitoredClusterIds: ValueBox[Set[Long]] = ValueBox(Set.empty)
```

File: http/src/main/resources/swagger/api-docs.yaml [hunk 18]
```diff
@@ -986,6 +986,9 @@ definitions:
         items:
           type: string
         description: The scopes for the cluster.
+      customClusterEnvironmentVariables:
```
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml:989]: is this change related to this PR?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 19]
```diff
@@ -190,134 +159,48 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
                                                                  cluster.clusterName)
       _ <- hasPermission match {
         case false =>
-          IO(
-            logger.warn(
+          log
```
Comment (kyuksel) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:161]: wow..scalafmt took it to another level here 🙈

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 20]
```diff
@@ -446,12 +454,17 @@ trait ClusterComponent extends LeoComponent {
...
+          cluster.dataprocInfo.map(_.googleId),
+          cluster.dataprocInfo.map(_.operationName.value),
+          cluster.dataprocInfo.map(_.stagingBucket.value),
           Timestamp.from(Instant.now)
         )
 
+    def clearAsyncClusterCreationFields(cluster: Cluster): DBIO[Int] =
+      findByIdQuery(cluster.id)
+        .map(c => (c.initBucket, c.serviceAccountKeyId, c.googleId, c.operationName, c.stagingBucket, c.dateAccessed))
+        .update(None, None, None, None, None, Timestamp.from(Instant.now))
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: `Instant.now` is side effecting..So ideally it should be passed in as parameter

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala (hunk 17 above)
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala:115]: should this be `ClusterInternalId`? since we've changed clusterId to uuid recently

File: http/src/main/resources/swagger/api-docs.yaml (hunk 18 above)
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml:989]: This field was added in a previous PR, but I noticed it was not added to swagger. I also had to persist it in the database in this PR, since back Leo is doing the cluster creation.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 20 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Yeah.. or use a DB function like `current_timestamp`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala (hunk 17 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala:115]: `ClusterInternalId` is a UUID, but we still have `id` which is a `Long`. Either should work here..

File: http/src/main/resources/reference.conf (hunk 15 above)
Comment (rtitle) [reply http/src/main/resources/reference.conf:109]: No, but I can create a wiki page..

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala (hunk 16 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala:197]: Because `Error` is an unrecoverable status. We don't want a user to start their cluster; have it time out; then have Leo error out (and delete) their cluster, potentially losing data.

This code transitions the cluster back to `Stopped`, so the user can try again or create a ticket -- either way, we won't delete their data.

File: http/src/main/resources/reference.conf (hunk 15 above)
Comment (kyuksel) [reply http/src/main/resources/reference.conf:109]: Either that (with a link to it here) or we could add that info directly here

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala (hunk 16 above)
Comment (kyuksel) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala:197]: got it

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala (hunk 17 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala:115]: I think this is safe -- this is an actor, and changes to `monitoredClusterIds` is only done in the actor context (not by a future). Actors provide a safe encapsulation of state, so mutable `var`s are allowed. I agree it's probably easy to access this `var` in an unsafe way, which is one of the things that bothers me about akka..

File: http/src/main/resources/reference.conf (hunk 15 above)
Comment (rtitle) [reply http/src/main/resources/reference.conf:109]: Good idea, will add here.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 20 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Slick apparently doesn't support updates with non-"pure" values. So you can't do:
```
update TABLE set col = current_timestamp
```
I instead changed these methods to take an `Instant` as a parameter, and had callers do stuff like:
```
for {
  now <- IO(Instant.now)
  _ <- dbRef.inTransactionIO { _.clusterQuery. clearAsyncClusterCreationFields(cluster, now) }
} yield ()
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 21]
```diff
@@ -1121,183 +645,54 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
+            log.error(e)(
               s"User ${userEmail.value} does not have access to ${gcsPath.bucketName} / ${gcsPath.objectName}"
-            )
-            throw BucketObjectAccessException(userEmail, gcsPath)
-          case e if whenGoogle401(e) =>
-            logger.warn(s"Could not validate object [${gcsUri}] as user [${userEmail.value}]", e)
-            ()
+            ) >> IO.raiseError(BucketObjectAccessException(userEmail, gcsPath))
+          case e if when401(e) =>
+            log.warn(e)(s"Could not validate object [${gcsUri}] as user [${userEmail.value}]") >> IO.unit
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: `IO.unit` is not needed since `log.warn` returns `IO.unit` I think

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala [hunk 22]
```diff
@@ -76,110 +61,55 @@ class BucketHelper(
...
-   */
-  def updateUserBucket(bucketName: GcsBucketName,
-                       googleProject: GoogleProject,
-                       serviceAccountInfo: ServiceAccountInfo): Future[Unit] =
-    for {
-      // available service accounts ((cluster or default SA) and notebook SA, if they exist) -> Reader
-      bucketSAs <- getBucketSAs(googleProject, serviceAccountInfo)
-
-      _ <- setBucketAcls(bucketName, bucketSAs, List.empty)
+      providerReaders <- Stream.eval(
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:71]: just want to point out, that if `providerReaders` is empty, then everything after this line won't happen. Is this intended with `insertBucket` and `setIamPolicy`?

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:71]: Hmm… good catch, this is probably not intended. We should be `++`ing the readers and groups. Will fix.

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:71]: Wait.. am I missing something?
```
providerReaders <- Stream.eval(serviceAccountProvider.listUsersStagingBucketReaders(userEmail).map(_.map(userIdentity)))
providerGroups <- Stream.eval(serviceAccountProvider.listGroupsStagingBucketReaders(userEmail).map(_.map(groupIdentity)))
```
These would still evaluate if `providerReaders` and `providerGroups` are empty right? The stream always contains 1 element, even if that element is `List.empty`.

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:71]:
```
scala> for {
     | a <- Stream.eval(IO(List.empty))
     | b <- Stream.eval(IO(List(1,2,3)))
     | } yield a ++ b
res9: fs2.Stream[[x]cats.effect.IO[x],List[Int]] = Stream(..)

scala> res9.compile.last.unsafeRunSync
res10: Option[List[Int]] = Some(List(1, 2, 3))
```

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:71]: oh my bad, I think this is fine...I thought `Stream.eval(serviceAccountProvider.listUsersStagingBucketReaders(userEmail).map(_.map(userIdentity)))` is creating a `Stream[IO, someReader]`, but it's actually `Stream[IO, List[someReader]]`

I was thinking something like
```
for {
    a <- Stream.emits(List.empty[Int]).covary[IO]
    b <- Stream.eval(IO(List(1,2,3)))
  } yield a :: b
res4: Stream[IO[x], List[Int]] = Stream(..)

@ res4.compile.toList.unsafeRunSync
res5: List[List[Int]] = List()
```

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BucketHelper.scala:71]: Right.. that makes sense
